README.md
[![Build Status](https://travis-ci.org/sixhours-team/memcached-spring-boot.svg?branch=master)](https://travis-ci.org/sixhours-team/memcached-spring-boot) 
[![codecov](https://codecov.io/gh/sixhours-team/memcached-spring-boot/branch/master/graph/badge.svg)](https://codecov.io/gh/sixhours-team/memcached-spring-boot)
[![Quality Gate](https://sonarcloud.io/api/project_badges/measure?project=io.sixhours%3Amemcached-spring-boot-autoconfigure&metric=alert_status)](https://sonarcloud.io/dashboard/index/io.sixhours%3Amemcached-spring-boot-autoconfigure)
[![Join the chat at gitter.im/six-hours/memcached-spring-boot](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/six-hours/memcached-spring-boot?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](http://www.apache.org/licenses/LICENSE-2.0.html)

# Memcached Spring Boot

Library that provides support for auto-configuration of Memcached cache in a Spring Boot application.

It provides implementation for the [Spring Cache Abstraction](https://docs.spring.io/spring/docs/4.3.x/spring-framework-reference/html/cache.html), backed by the [Amazon's ElastiCache Clustered Client](https://github.com/awslabs/aws-elasticache-cluster-client-memcached-for-java).
Supports cache eviction per key, as well as clearing out of the entire cache region. Binaries are available from **Maven Central** and **JCenter**.

## Usage

To plug-in Memcached cache in your application follow the steps below:

1. Include library as a Gradle or Maven compile dependency:
   * **Gradle**
   
      ```groovy
      compile('io.sixhours:memcached-spring-boot-starter:1.3.1') 
      ```
   * **Maven**
   
      ```xml
      <dependency>
          <groupId>io.sixhours</groupId>
          <artifactId>memcached-spring-boot-starter</artifactId>
          <version>1.3.1</version>
      </dependency>
      ```
2. Configure `Memcached` key-value store in your properties file (`application.yml`).

    **Example**

    To manually connect to one or more cache servers (nodes), specify comma-separated list of hostname:port with the `static` mode:
       
    ```yaml
     memcached.cache:
       servers: example1.com:11211,example2.com:11211
       mode: static
       expirations: 86400, cache_name1:3600, cache_name2:108000 # global expiration is '86400' and custom ones for cache_name1 and cache_name2
     ```

    To connect to a cluster with AWS [Auto Discovery](http://docs.aws.amazon.com/AmazonElastiCache/latest/UserGuide/AutoDiscovery.html), specify
    cluster configuration endpoint in `memcached.cache.servers` property with the `dynamic` mode:
   
    ```yaml
    memcached.cache:
        servers: mycluster.example.com:11211
        mode: dynamic
        expirations: 86400 # global expiration set to '86400'
    ```
   
3. Enable caching support by adding `@EnableCaching` annotation to one of your `@Configuration` classes.

    ```java
    import org.springframework.boot.SpringApplication;
    import org.springframework.boot.autoconfigure.SpringBootApplication;
    import org.springframework.cache.annotation.EnableCaching;
    
    @SpringBootApplication
    @EnableCaching
    public class Application {
    
    	public static void main(String[] args) {
    		SpringApplication.run(Application.class, args);
    	}
    }
    ```

    Now you can add caching to an operation of your service:
 
    ```java
    import org.springframework.cache.annotation.Cacheable;
    import org.springframework.stereotype.Component;
    
    @Component
    public class BookService {
    
        @Cacheable("books")
        public Book findByTitle(String title) {
            // ...
        }
    
    }
    ```

For further details on using the Memcached cache in a Spring Boot application please look at the [demo](https://github.com/sixhours-team/spring-boot-memcached-demo-java) project. 

## Properties

Properties can be set in your `application.yml` file or as a command line properties. Below is the
full list of supported properties:

```yaml
# MEMCACHED CACHE 
memcached.cache.servers: # Comma-separated list of hostname:port for memcached servers (default "localhost:11211")
memcached.cache.mode: # Memcached client mode (use one of following: "static", "dynamic"). Default mode is "static", use "dynamic" for AWS node auto discovery
memcached.cache.expirations: # Cache expirations in seconds (default "60"). To set new global expiration use value without colon: {number} e.g. "86400". To set value per cache name use format: {cache_name}:{number} e.g. "authors:3600"
memcached.cache.prefix: # Cache key prefix (default "memcached:spring-boot")
memcached.cache.namespace: # Cache eviction namespace key name (default "namespace")
memcached.cache.protocol: # Memcached client protocol. Supports "text" and "binary" protocols (default is "text" protocol)
```

All of the values have sensible defaults and are bound to [MemcachedCacheProperties](https://github.com/sixhours-team/memcached-spring-boot/blob/master/memcached-spring-boot-autoconfigure/src/main/java/io/sixhours/memcached/cache/MemcachedCacheProperties.java) class.

**Notice:** 
>If different applications are sharing the same Memcached server, make sure to specify unique cache `prefix` for each application 
in order to avoid cache conflicts.

## Build

In order to build the project you will have to have [Java 1.8+](http://www.oracle.com/technetwork/java/javase/downloads/index.html) and [Docker](https://www.docker.com/get-docker) installed.
To build the project invoke the following command:

    ./gradlew clean build
    
To install the modules in the local Maven repository:

    ./gradlew clean build install
    
## License

Memcached Spring Boot is an Open Source software released under the [Apache 2.0 license](http://www.apache.org/licenses/LICENSE-2.0.html).


CHANGELOG.md
# Memcached Spring Boot Releases #

### Version 1.3.1 ###

* [Pull 47](https://github.com/sixhours-team/memcached-spring-boot/pull/47) Reset namespace key expiration on every new value addition to the same cache region
* [Pull 45](https://github.com/sixhours-team/memcached-spring-boot/pull/45) Upgrade gradle wrapper to 4.10.2
* [Pull 43](https://github.com/sixhours-team/memcached-spring-boot/pull/43) Upgrade to Gradle 4.7 
* [Pull 42](https://github.com/sixhours-team/memcached-spring-boot/pull/42) Fix SonarCloud status display in README.md 

### Version 1.3.0 ###

* [Pull 41](https://github.com/sixhours-team/memcached-spring-boot/pull/41) Expiration per cache name

### Version 1.2.0 ###

* [Pull 39](https://github.com/sixhours-team/memcached-spring-boot/pull/39) Update gradle wrapper to 4.6
* [Pull 38](https://github.com/sixhours-team/memcached-spring-boot/pull/38) Add author service integration test
* [Pull 36](https://github.com/sixhours-team/memcached-spring-boot/pull/36) Add support for Spring Cloud refresh scope beans
* [Pull 34](https://github.com/sixhours-team/memcached-spring-boot/pull/34) Add deprecated javadoc for namespace configuration
* [Pull 33](https://github.com/sixhours-team/memcached-spring-boot/pull/33) Add name, description and url to the generated pom files

### Version 1.1.0 ###

* [Pull 32](https://github.com/sixhours-team/memcached-spring-boot/pull/32) Deprecate namespace configuration property; use default value instead
* [Pull 31](https://github.com/sixhours-team/memcached-spring-boot/pull/31) Update springboot version to 1.5.10
* [Pull 29](https://github.com/sixhours-team/memcached-spring-boot/pull/29) Add client protocol configuration property
* [Pull 28](https://github.com/sixhours-team/memcached-spring-boot/pull/28) Add author service integration test
* [Pull 27](https://github.com/sixhours-team/memcached-spring-boot/pull/27) Upgrade gradle wrapper to 4.5
* [Pull 26](https://github.com/sixhours-team/memcached-spring-boot/pull/26) Replace mvn publish with mvn plugin to fix optional scope in pom file      

### Version 1.0.2 ###

* [Pull 24](https://github.com/sixhours-team/memcached-spring-boot/pull/24) Add project key for Sonar Cloud
* [Pull 23](https://github.com/sixhours-team/memcached-spring-boot/pull/23) Bintray organization
* [Pull 22](https://github.com/sixhours-team/memcached-spring-boot/pull/22) Make server list read-only
* [Pull 21](https://github.com/sixhours-team/memcached-spring-boot/pull/21) Sonarqube
* [Pull 20](https://github.com/sixhours-team/memcached-spring-boot/pull/20) Store null values
* [Pull 19](https://github.com/sixhours-team/memcached-spring-boot/pull/19) Initialize namespace with the timestamp
* [Pull 18](https://github.com/sixhours-team/memcached-spring-boot/pull/18) Update test runner
* [Pull 17](https://github.com/sixhours-team/memcached-spring-boot/pull/17) Increase cache expiration in tests
* [Pull 16](https://github.com/sixhours-team/memcached-spring-boot/pull/16) IntelliJ testsource configuration
* [Pull 15](https://github.com/sixhours-team/memcached-spring-boot/pull/15) Add license info for test classes
* [Pull 14](https://github.com/sixhours-team/memcached-spring-boot/pull/14) Use AssertJ for test results verification
* [Pull 13](https://github.com/sixhours-team/memcached-spring-boot/pull/13) Update integration tests
* [Pull 12](https://github.com/sixhours-team/memcached-spring-boot/pull/12) Add integration tests
* [Pull 11](https://github.com/sixhours-team/memcached-spring-boot/pull/11) Update links to organization url
* [Pull 10](https://github.com/sixhours-team/memcached-spring-boot/pull/10) Use InetSocketAddress for server list
* [Pull 9](https://github.com/sixhours-team/memcached-spring-boot/pull/9) Lombok removed from project

### Version 1.0.1 ###

* [Pull 5](https://github.com/sixhours-team/memcached-spring-boot/pull/5) The library should honour `spring.cache.type` property [Issue 4](https://github.com/sixhours-team/memcached-spring-boot/issues/4) 

### Version 1.0.0 ###

* Initial release


memcached-spring-boot-autoconfigure/src/main/java/io/sixhours/memcached/cache/Default.java
/*
 * Copyright 2017 Sixhours.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.sixhours.memcached.cache;

import net.spy.memcached.ClientMode;

import java.net.InetSocketAddress;
import java.util.Collections;
import java.util.List;

import static io.sixhours.memcached.cache.MemcachedCacheProperties.Protocol;

/**
 * Default cache configuration values.
 *
 * @author Igor Bolic
 */
public final class Default {

    public static final List<InetSocketAddress> SERVERS = Collections.unmodifiableList(
            Collections.singletonList(new InetSocketAddress("localhost", 11211)));

    public static final ClientMode CLIENT_MODE = ClientMode.Static;

    public static final int EXPIRATION = 60;

    public static final String PREFIX = "memcached:spring-boot";

    public static final String NAMESPACE = "namespace";

    public static final Protocol PROTOCOL = Protocol.TEXT;

    private Default() {
        throw new AssertionError("Suppress default constructor");
    }
}
